File: web/js/peacoq-ts/goals.ts
import { Goal } from "./goal";

export class Goals implements IGoals {
  fgGoals: Goal[];
  bgGoals: GoalBeforeAfter[];
  shelvedGoals: Goal[];
  givenUpGoals: Goal[];

  constructor(maybeGoals: any) {
    if (!maybeGoals) {
      this.fgGoals = [];
      this.bgGoals = [];
      this.shelvedGoals = [];
      this.givenUpGoals = [];
    } else {
      this.fgGoals = _(maybeGoals[0]).map(function(g: any) {
        return new Goal(g);
      }).value();
      this.bgGoals = _(maybeGoals[1]).map(function(ba: any) {
        return {
          "before": _(ba[0]).map(function(b: any) { return new Goal(b); }).value(),
          "after": _(ba[1]).map(function(b: any) { return new Goal(b); }).value(),
        };
      }).value();
      this.shelvedGoals = _(maybeGoals[2]).map(function(g: any) {
        return new Goal(g);
      }).value();
      this.givenUpGoals = _(maybeGoals[3]).map(function(g: any) {
        return new Goal(g);
      }).value();
    }
  }

}

File: src/lib/Coq/Goals.hs
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE OverloadedStrings #-}

module Coq.Goals where

import Data.Aeson            (ToJSON)
import Data.Proxy
import GHC.Generics          (Generic)
import Text.XML.Stream.Parse (tagNoAttr)

import Coq.Goal
import FromXML

data PreGoal a
  = MkPreGoal [a] [([a], [a])] [a] [a]
  deriving (Generic, Show)

instance ToJSON a => ToJSON (PreGoal a)

data Goals
  = MkGoals (PreGoal Goal)
  deriving (Generic, Show)

instance ToJSON Goals

instance FromXML Goals where
  instanceName Proxy = "Goals"
  parseXML = tagNoAttr "goals" $ do
    fgGoals <- forceXML
    bgGoals <- forceXML
    shelvedGoals <- forceXML
    givenUpGoals <- forceXML
    return $ MkGoals (MkPreGoal fgGoals bgGoals shelvedGoals givenUpGoals)
